### PR TITLE
WIP: Tests timeouts and change prestarts to the new structure of the other PR

### PR DESCRIPTION
**Pull request by lfdversluis**
This PR is meant to gather some feedback about ongoing changes. In the other PR which fixed the race condition, the  ``prestart`` function is gone, so fixing that as tests are broken now. Also adding timeotus when proxies etc. cannot be created.

### CODE REVIEW

File: Tribler/Main/metadata-injector.py [hunk 1]
```diff
@@ -81,7 +82,13 @@ def initialize(self):
...
         self._logger.info(u"Starting session...")
         self.session = Session(sscfg)
-        self.session.prestart()
+        self.session.initialize_database()
+
+        failed, has_to_upgrade = self.session.has_to_upgrade_database()
+        if has_to_upgrade:
+            self.session.upgrade_database()
+        elif failed:
+            self.session.stash_database()
```
Comment (lfdversluis) [inline Tribler/Main/metadata-injector.py]: Hmmm this piece of code is duplicate a few times. Maybe add create a function in session for it?

Comment (devos50) [inline Tribler/Main/metadata-injector.py]: Yes, a seperate method would be better indeed.

File: Tribler/Test/test_as_server.py [hunk 2]
```diff
@@ -353,6 +363,8 @@ def setUp(self):
         self.asserts = []
         self.annotate(self._testMethodName, start=True)
 
+        print "\nHERE"
```
Comment (devos50) [inline Tribler/Test/test_as_server.py]: This print statement should be removed.

File: Tribler/Test/test_as_server.py [hunk 3]
```diff
@@ -390,6 +402,8 @@ def startTest(self, callback, min_callback_delay=5, autoload_discovery=True,
         from Tribler.Main.vwxGUI.GuiUtility import GUIUtility
         from Tribler.Main import tribler_main
 
+        print "\nHERE2"
```
Comment (devos50) [inline Tribler/Test/test_as_server.py]: This statement should also be removed when ready.

**Comment by devos50** [comment]
The PR contains all changes that were also proposed in #1756. Maybe we should merge this PR with #1756 to have one set of changes?

**Comment by lfdversluis** [comment]
@whirm likes stuff to be separated. I indeed branched of that branch because it contained critical fixes i.e. fix segfaults. That one should be merged first indeed, then I'll rebase and see what happens

File: Tribler/Test/test_tunnel_base.py [hunk 4]
```diff
@@ -19,18 +21,34 @@ class TestTunnelBase(TestGuiAsServer):
...
+
         def setup_proxies():
+            print "\nHERE12"
             tunnel_communities = []
             baseindex = 3
             for i in range(baseindex, baseindex + nr_relays):  # Normal relays
-                tunnel_communities.append(create_proxy(i, False, crypto_enabled))
+                self.CallConditional(60, create_proxy(i, False, crypto_enabled), print_proxy_setup_succsesful(False, i),
+                                 'Could not create normal proxy within 60 seconds')
+                # tunnel_communities.append(create_proxy(i, False, crypto_enabled))
```
Comment (lfdversluis) [inline Tribler/Test/test_tunnel_base.py]: This section is where the test currently hangs because of a lack of proxies and/or exit nodes.

**Comment by lfdversluis** [comment]
@whirm This contains the changes in #1756 But as this PR also contains the `prestart` replacements. Is it OK to close the other PR and contain all the fixes in this one. I know this PR is rather big then but #1756 kind of depends on this one.

**Comment by whirm** [comment]
Do the fixes depend on each other?

**Comment by lfdversluis** [comment]
@whirm Yes the dialog fix replaces the prestart, but to fix the test (which makes use of prestart in the code) I need dialog fix because it fixes the seg fault and race condition :smile:

**Comment by whirm** [comment]
Ok, then let's do everything on a single one, you can close the other while I review this one.

**Comment by lfdversluis** [comment]
I am going to refactor the code duplication (check and if needed upgrade / stash db). I'll make a function in session

File: Tribler/Core/Upgrade/upgrade.py [hunk 5]
```diff
@@ -20,10 +19,24 @@
 
 class TriblerUpgrader(object):
 
-    def __init__(self, session, db):
+    _singleton = None
```
Comment (whirm) [inline Tribler/Core/Upgrade/upgrade.py]: We don't want any more singletons, it makes things harder to test.

Just make each Session instance its own Upgrade object.

File: Tribler/Core/Session.py [hunk 6]
```diff
@@ -152,19 +150,39 @@ def set_and_create_dir(dirname, setter, default_dir):
...
         """
-        # initialize the database
         self.sqlite_db = SQLiteCacheDB(self)
         self.sqlite_db.initialize()
         self.sqlite_db.initial_begin()
 
-        # check and upgrade
-        upgrader = TriblerUpgrader(self, self.sqlite_db)
-        upgrader.check_and_upgrade()
-        return upgrader
```
Comment (whirm) [inline Tribler/Core/Session.py]: These are just wrappers for the Upgrade object, why aren't you just using these directly instead?

**Comment by synctext** [comment]
what is the 'stash db' functionality?

**Comment by whirm** [comment]
@synctext it moves the old and non-compatible state dir away so the new Tribler can start from scratch.

**Comment by lfdversluis** [comment]
retest this please

**Comment by lfdversluis** [comment]
Gonna close this one and open several smaller PRs

**Pull request closed without merging**